File: pty.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"io"
	"net"
	"os"
	"path"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	taskapi "github.com/containerd/containerd/api/runtime/task/v3"
	"github.com/containerd/containerd/v2/pkg/namespaces"
	"github.com/containerd/errdefs"
	"github.com/containerd/errdefs/pkg/errgrpc"
	"github.com/containerd/log"
	systemd "github.com/coreos/go-systemd/v22/dbus"
	dbus "github.com/godbus/dbus/v5"
	"github.com/sirupsen/logrus"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"google.golang.org/protobuf/types/known/emptypb"
)

const (
	ttySockPathEnv  = "_TTY_SOCKET_PATH"
	ttyHandshakeEnv = "_TTY_HANDSHAKE"
)

// The tty handler protocol is a small newline-framed, text request/response
// exchanged over a persistent unix socket. A request is "<op> <width> <height>"
// and a response is a single status line: "0" on success or "1 <message>" on
// error. These constants are mirrored by op_resize/reply_* in pty.c.
const (
	ttyOpResize  = 1
	ttyStatusOK  = "0"
	ttyStatusErr = "1"

	// maxTTYResponseLen bounds how many bytes we buffer for a single response
	// line so a misbehaving handler cannot make us read without limit. It
	// comfortably exceeds the longest reply pty.c can produce.
	maxTTYResponseLen = 256

	// ttyRequestBufSize sizes the reusable request buffer. A resize request is
	// "<op> <width> <height>\n"; width and height arrive as uint32 (at most 10
	// digits each), so the longest possible request, "1 4294967295 4294967295\n",
	// is 24 bytes. Rounded up for headroom.
	ttyRequestBufSize = 32

	// ttyResponseTimeout bounds a single resize exchange. The read runs while the
	// process lock is held, so without a deadline a stuck handler -- or an old,
	// pre-framing handler that writes a bare "0" and never sends the newline this
	// client now waits for -- would hang the whole lifecycle. The timeout is
	// generous: a healthy handler replies in microseconds.
	ttyResponseTimeout = 2 * time.Second
)

// ttyControlClient sends control operations to a container's tty handler over
// its unix control socket. The handler speaks the small newline-framed
// request/response protocol described above; this client hides that behind
// Resize and Close, dialing lazily and reusing the connection across resizes.
// Its own mutex serializes exchanges, so tty I/O never blocks the process
// lifecycle lock.
type ttyControlClient struct {
	sockPath string

	mu     sync.Mutex
	conn   net.Conn
	reader *bufio.Reader
	limit  *io.LimitedReader
	// buf is a reusable scratch buffer for encoding each request, so a resize
	// does not allocate.
	buf [ttyRequestBufSize]byte
}

func newTTYControlClient(sockPath string) *ttyControlClient {
	return &ttyControlClient{sockPath: sockPath}
}

// Resize sets the tty window size. It dials the handler on first use and reuses
// the connection thereafter; if the exchange fails it drops the connection so
// the next resize redials.
func (c *ttyControlClient) Resize(ctx context.Context, width, height int) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn == nil {
		if err := c.dial(); err != nil {
			return err
		}
	}

	line, err := c.exchange(ctx, appendTTYResizeRequest(c.buf[:0], width, height))
	if err != nil {
		// The handler is a transient local unit: a failed exchange means it is
		// gone, not that a retry would help, so just drop the connection and let
		// the next resize redial.
		c.reset()
		return err
	}
	return parseTTYResponse(line)
}

// Close closes the control connection. It is safe to call more than once and
// concurrently with Resize. The caller must not hold c.mu.
func (c *ttyControlClient) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.reset()
}

// dial establishes the control connection and its bounded reader. The caller
// holds c.mu.
func (c *ttyControlClient) dial() error {
	conn, err := net.Dial("unix", c.sockPath)
	if err != nil {
		return fmt.Errorf("could not dial tty sock: %w", err)
	}
	c.conn = conn
	c.limit = &io.LimitedReader{R: conn}
	c.reader = bufio.NewReader(c.limit)
	return nil
}

// reset closes and clears the control connection, returning the close error.
// The caller holds c.mu.
func (c *ttyControlClient) reset() error {
	if c.conn == nil {
		return nil
	}
	err := c.conn.Close()
	c.conn, c.reader, c.limit = nil, nil, nil
	return err
}

// exchange writes req and reads one response line, bounded by ttyResponseTimeout
// (or an earlier context deadline). The caller holds c.mu.
func (c *ttyControlClient) exchange(ctx context.Context, req []byte) (string, error) {
	deadline := time.Now().Add(ttyResponseTimeout)
	if d, ok := ctx.Deadline(); ok && d.Before(deadline) {
		deadline = d
	}
	if err := c.conn.SetDeadline(deadline); err != nil {
		return "", fmt.Errorf("error setting tty handler deadline: %w", err)
	}

	if _, err := c.conn.Write(req); err != nil {
		return "", fmt.Errorf("error writing winsize to the tty handler: %w", err)
	}

	line, err := c.readMsg()
	if err != nil {
		return "", fmt.Errorf("error reading response from tty handler: %w", err)
	}
	return line, nil
}

// readMsg reads one status line from the handler, bounding the read at
// maxTTYResponseLen bytes via the reusable limit. A well-behaved handler ends
// the line with a newline. An older, pre-framing handler writes an unframed
// reply (e.g. a bare "0") then waits for the next request, so no newline
// arrives; the read then ends by deadline or EOF with the message buffered, and
// we return it rather than hanging or reporting a failure. The caller holds c.mu.
func (c *ttyControlClient) readMsg() (string, error) {
	c.limit.N = maxTTYResponseLen
	line, err := c.reader.ReadString('\n')
	if err != nil {
		if c.limit.N <= 0 {
			return "", fmt.Errorf("tty handler response exceeded %d bytes without a newline", maxTTYResponseLen)
		}
		if len(line) == 0 {
			return "", err
		}
		// Unframed reply: the message is buffered but no newline arrived.
	}
	return strings.TrimRight(line, "\r\n"), nil
}

// appendTTYResizeRequest appends a newline-terminated resize request to buf and
// returns the extended buffer, so a caller can reuse a single buffer across
// resizes instead of allocating one per call.
func appendTTYResizeRequest(buf []byte, width, height int) []byte {
	buf = strconv.AppendInt(buf, ttyOpResize, 10)
	buf = append(buf, ' ')
	buf = strconv.AppendInt(buf, int64(width), 10)
	buf = append(buf, ' ')
	buf = strconv.AppendInt(buf, int64(height), 10)
	return append(buf, '\n')
}

// parseTTYResponse interprets a single status line from the tty handler.
func parseTTYResponse(line string) error {
	line = strings.TrimRight(line, "\r\n")
	switch {
	case line == ttyStatusOK:
		return nil
	case line == ttyStatusErr:
		return fmt.Errorf("tty handler returned an unspecified error")
	case strings.HasPrefix(line, ttyStatusErr+" "):
		return fmt.Errorf("tty handler returned an error: %s", strings.TrimPrefix(line, ttyStatusErr+" "))
	default:
		return fmt.Errorf("tty handler returned an unexpected response: %q", line)
	}
}

func (p *process) ResizePTY(ctx context.Context, width, height int, sockPath string) error {
	if !p.Terminal {
		// This mimics what the runc shim does, and what the containerd integration tests expect
		return nil
	}

	p.mu.Lock()
	if p.ttyControl == nil {
		p.ttyControl = newTTYControlClient(sockPath)
	}
	c := p.ttyControl
	p.mu.Unlock()

	return c.Resize(ctx, width, height)
}

// closeTTYControl detaches and closes the process's tty control client, if any.
// It holds p.mu only to detach the pointer, never across the close, so a slow or
// stuck resize cannot wedge the lifecycle lock.
func (p *process) closeTTYControl() {
	p.mu.Lock()
	c := p.ttyControl
	p.ttyControl = nil
	p.mu.Unlock()
	if c != nil {
		c.Close()
	}
}

// ResizePty of a process
func (s *Service) ResizePty(ctx context.Context, r *taskapi.ResizePtyRequest) (_ *emptypb.Empty, retErr error) {
	ns, err := namespaces.NamespaceRequired(ctx)
	if err != nil {
		return nil, err
	}

	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(logrus.Fields{
		"id":        r.ID,
		"ns":        ns,
		"apiAction": "resizePty",
	}))

	log.G(ctx).Debug("systemd.ResizePTY start")
	defer func() {
		log.G(ctx).WithError(retErr).Debug("systemd.ResizePTY end")
		if retErr != nil {
			retErr = errgrpc.ToGRPC(retErr)
		}
	}()

	p := s.processes.Get(path.Join(ns, r.ID))
	if p == nil {
		return nil, fmt.Errorf("process %s: %w", r.ID, errdefs.ErrNotFound)
	}

	if r.ExecID != "" {
		ep := p.(*initProcess).execs.Get(r.ExecID)
		if err := ep.ResizePTY(ctx, int(r.Width), int(r.Height)); err != nil {
			return nil, err
		}
	} else {
		if err := p.ResizePTY(ctx, int(r.Width), int(r.Height)); err != nil {
			return nil, err
		}
	}

	return &emptypb.Empty{}, nil
}

func (p *execProcess) ResizePTY(ctx context.Context, w, h int) error {
	ttyPath, err := p.ttySockPath()
	if err != nil {
		return err
	}
	return p.process.ResizePTY(ctx, w, h, ttyPath)
}

func (p *initProcess) ResizePTY(ctx context.Context, w, h int) error {
	ttyPath, err := p.ttySockPath()
	if err != nil {
		return err
	}
	return p.process.ResizePTY(ctx, w, h, ttyPath)
}

// CloseIO of a process
func (s *Service) CloseIO(ctx context.Context, r *taskapi.CloseIORequest) (_ *emptypb.Empty, retErr error) {
	// TODO: I'm not sure what we should do here since we aren't really doing anything with container I/O
	// Potentially we should signal the tty handler to stop copying stdio?
	return &emptypb.Empty{}, nil
}

func (p *initProcess) ttySockPath() (string, error) {
	sockInfoPath := filepath.Join(p.root, "tty.sock")
	b, err := os.ReadFile(sockInfoPath)
	if err == nil {
		return string(b), nil
	}

	if !os.IsNotExist(err) {
		return "", err
	}

	tmp, err := os.MkdirTemp(os.Getenv("XDG_RUNTIME_DIR"), "pty")
	if err != nil {
		return "", err
	}
	s := filepath.Join(tmp, "s")
	if err := os.WriteFile(sockInfoPath, []byte(s), 0600); err != nil {
		os.RemoveAll(tmp)
		return "", err
	}

	return s, nil
}

func (p *execProcess) ttySockPath() (string, error) {
	sockInfoPath := filepath.Join(p.stateDir(), "tty.sock")
	b, err := os.ReadFile(sockInfoPath)
	if err == nil {
		return string(b), nil
	}

	if !os.IsNotExist(err) {
		return "", err
	}

	tmp, err := os.MkdirTemp(os.Getenv("XDG_RUNTIME_DIR"), "pty")
	if err != nil {
		return "", err
	}
	s := filepath.Join(tmp, "s")
	if err := os.WriteFile(sockInfoPath, []byte(s), 0600); err != nil {
		os.RemoveAll(tmp)
		return "", err
	}

	return s, nil
}

func (p *process) ttyUnitName() string {
	return unitName(p.ns, p.id, "tty")
}

func (p *process) makePty(ctx context.Context, sockPath string) (_, _ string, retErr error) {
	ctx, span := StartSpan(ctx, "process.StartTTY")
	defer func() {
		if retErr != nil {
			span.SetStatus(codes.Error, retErr.Error())
		}
		span.End()
	}()

	logPath := filepath.Join(p.root, p.id+"-tty.log")
	defer func() {
		if retErr != nil {
			logData, err := os.ReadFile(logPath)
			if err != nil {
				log.G(ctx).WithError(err).Info("Could not read tty error log")
			} else {
				retErr = fmt.Errorf("%w: %s", retErr, string(logData))
			}
		}
	}()

	if p.Stdin != "" {
		f, _ := os.OpenFile(p.Stdin, os.O_RDWR, 0)
		if f != nil {
			defer f.Close()
		}
	}

	if p.Stdout != "" {
		f, _ := os.OpenFile(p.Stdout, os.O_RDWR, 0)
		if f != nil {
			defer f.Close()
		}
	}

	if p.Stderr != "" {
		f, _ := os.OpenFile(p.Stderr, os.O_RDWR, 0)
		if f != nil {
			defer f.Close()
		}
	}
	properties := p.ttyServiceProperties(sockPath, logPath)

	ttyUnit := p.ttyUnitName()
	defer func() {
		if retErr != nil {
			p.systemd.StopUnitContext(ctx, ttyUnit, "replace", nil)
		}
	}()

	chTTY := make(chan string, 1)
	if _, err := p.systemd.StartTransientUnitContext(ctx, ttyUnit, "replace", properties, chTTY); err != nil {
		if e := p.systemd.ResetFailedUnitContext(ctx, ttyUnit); e == nil {
			_, err2 := p.systemd.StartTransientUnitContext(ctx, ttyUnit, "replace", properties, chTTY)
			if err2 == nil {
				err = nil
			}
		} else {
			log.G(ctx).WithField("unit", ttyUnit).WithError(e).Warn("Error reseting failed unit")
		}
		if err != nil {
			return "", "", fmt.Errorf("error starting tty service: %w", err)
		}
	}
	select {
	case <-ctx.Done():
	case status := <-chTTY:
		span.SetAttributes(attribute.String("tty-unit-status", status))
		if status != "done" {
			return "", "", fmt.Errorf("failed to start tty service: %s", status)
		}
	}

	return ttyUnit, sockPath, nil
}

func (p *process) ttyServiceProperties(sockPath, logPath string) []systemd.Property {
	env := []string{
		ttyHandshakeEnv + "=1",
		ttySockPathEnv + "=" + sockPath,
	}
	if p.shimCgroup != "" {
		env = append(env, "SHIM_CGROUP="+p.shimCgroup)
	}

	properties := []systemd.Property{
		systemd.PropType("notify"),
		systemd.PropExecStart([]string{p.exe, "tty-handshake"}, false),
		{Name: "Environment", Value: dbus.MakeVariant(env)},
		{Name: "StandardErrorFile", Value: dbus.MakeVariant(logPath)},
	}
	if p.Stdin != "" {
		properties = append(properties, systemd.Property{Name: "StandardInputFile", Value: dbus.MakeVariant(p.Stdin)})
	}
	if p.Stdout != "" {
		properties = append(properties, systemd.Property{Name: "StandardOutputFile", Value: dbus.MakeVariant(p.Stdout)})
	}
	return properties
}

File: pty.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>
#include <netdb.h>
#include <sys/un.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <termios.h>
#include <fcntl.h>
#include <libgen.h>

#include "systemd.h"
#include "log.h"

int op_resize = 1;
int sock_fd;
int tty_fd;

struct copy_data
{
    int w;
    int r;
};

void *copy(void *args)
{
    struct copy_data *cp;

    cp = (struct copy_data *)args;

    char buf[1024];
    int n;

    while ((n = read(cp->r, buf, sizeof(buf))) > 0)
        n = write(cp->w, buf, n);

    return 0;
}

// write_all writes the whole buffer, looping over short writes.
// Returns 0 on success and -1 if any write fails.
int write_all(int fd, const char *buf, size_t len)
{
    size_t off = 0;
    while (off < len)
    {
        ssize_t n = write(fd, buf + off, len - off);
        if (n < 0)
            return -1;
        off += (size_t)n;
    }
    return 0;
}

// The tty client protocol is a newline-framed status line per request:
//   "0\n"            on success
//   "1 <message>\n"  on error
// reply_ok/reply_err write one such line. They return 0 on success and -1 on
// write failure.
int reply_ok(int fd)
{
    return write_all(fd, "0\n", 2);
}

int reply_err(int fd, const char *msg)
{
    char buf[256];
    int n = snprintf(buf, sizeof(buf), "1 %s\n", msg);
    if (n < 0)
        return -1;
    if (n >= (int)sizeof(buf))
        n = sizeof(buf) - 1;
    return write_all(fd, buf, (size_t)n);
}

// handle_tty_request services a single request line (newline already stripped),
// writing exactly one status line back to the client. It returns 0 when the
// connection should stay open -- including recoverable errors that were reported
// to the client -- and -1 when a socket write failed and the caller should close.
int handle_tty_request(int fd, const char *line)
{
    int op, w, h;
    if (sscanf(line, "%d %d %d", &op, &w, &h) != 3)
    {
        lerror("parse");
        return reply_err(fd, "invalid resize request");
    }

    if (op != op_resize)
        return reply_err(fd, "unknown operation");

    struct winsize ws;
    memset(&ws, 0, sizeof(ws));
    ws.ws_col = w;
    ws.ws_row = h;

    if (ioctl(tty_fd, TIOCSWINSZ, &ws) < 0)
    {
        lerror("ioctl TIOCSWINSZ");
        return reply_err(fd, "error setting window size");
    }

    return reply_ok(fd);
}

void handle_tty_op_conn(int fd)
{
    char buf[256];
    // len is how many bytes of a not-yet-complete request are buffered. The
    // socket is a byte stream, so a single read() may return a partial request
    // or several requests at once; we buffer until each '\n' before parsing so
    // fragmentation cannot desynchronize requests from responses.
    size_t len = 0;

    while (1)
    {
        // Leave room to NUL-terminate for sscanf.
        ssize_t nr = read(fd, buf + len, sizeof(buf) - 1 - len);
        if (nr < 0)
        {
            lerror("read");
            close(fd);
            return;
        }
        if (nr == 0)
        {
            // The client closed the connection.
            close(fd);
            return;
        }
        len += (size_t)nr;

        // Service every complete, newline-terminated request in the buffer.
        char *start = buf;
        char *nl;
        while ((nl = memchr(start, '\n', (buf + len) - start)) != NULL)
        {
            *nl = '\0';
            if (handle_tty_request(fd, start) < 0)
            {
                lerror("write");
                close(fd);
                return;
            }
            start = nl + 1;
        }

        // Keep any trailing partial request for the next read.
        len -= (size_t)(start - buf);
        memmove(buf, start, len);

        // A request that fills the buffer without a newline is malformed.
        if (len == sizeof(buf) - 1)
        {
            reply_err(fd, "request too long");
            close(fd);
            return;
        }
    }
}

void *handle_tty_ops(void *args)
{
    // This fd came from go, so you know it is in non-blocking mode.
    // We aren't doing non-blocking I/O here so we need to set it back to blocking.
    int fd_flags = fcntl(sock_fd, F_GETFL);
    if (fd_flags < 0)
    {
        lerror("F_GETFL");
        return 0;
    }

    int err = fcntl(sock_fd, F_SETFL, fd_flags & ~O_NONBLOCK);
    if (err < 0)
    {
        lerror("F_SETFL");
        return 0;
    }

    lmsg("Ready for tty clients\n");

    while (1)
    {
        int cfd = accept(sock_fd, NULL, NULL);
        if (cfd < 0)
        {
            lerror("accept");
            close(sock_fd);
            return 0;
        }

        // Only handle one connection at a time.
        // This should only have one client.
        handle_tty_op_conn(cfd);
    }
}

void setcgroup(void)
{
    char *val = getenv("SHIM_CGROUP");
    if (val == NULL)
        return;

    char *CG_PATH = "/sys/fs/cgroup";
    char *CG_PROCS = "cgroup.procs";

    int sz = strlen(CG_PATH) + strlen(val) + 1 + strlen(CG_PROCS);
    char p[sz];

    sprintf(p, "%s/%s/%s", CG_PATH, val, CG_PROCS);
    lmsg("Setting cgroup");

    FILE *fd = fopen(p, "w");
    if (fd < 0)
    {
        lerror("fopen");
        return;
    }

    lmsg("opened cgroup.procs");

    pid_t pid = getpid();
    int err = fprintf(fd, "%d\n", pid);
    if (err < 0)
    {
        lerror("write cgroup.procs");
    }
    lmsg("wrote cgroup.procs");
    err = fclose(fd);
    if (err < 0)
    {
        lerror("fclose");
    }
}

int handle_pty(void)
{
    pthread_t stdin_copy_thr_id, stdout_copy_thr_id, tty_op_thr_id;
    struct copy_data *stdin_copy = (struct copy_data *)malloc(sizeof(struct copy_data));
    stdin_copy->w = tty_fd;
    stdin_copy->r = 0;
    pthread_create(&stdin_copy_thr_id, NULL, copy, (void *)stdin_copy);

    struct copy_data *stdout_copy = (struct copy_data *)malloc(sizeof(struct copy_data));
    stdout_copy->w = 1;
    stdout_copy->r = tty_fd;
    pthread_create(&stdout_copy_thr_id, NULL, copy, (void *)stdout_copy);

    pthread_create(&tty_op_thr_id, NULL, handle_tty_ops, (void *)NULL);

    pthread_join(stdout_copy_thr_id, NULL);
    free(stdout_copy);

    pthread_cancel(stdin_copy_thr_id);
    pthread_join(stdin_copy_thr_id, NULL);
    free(stdin_copy);

    close(tty_fd);
    close(sock_fd);

    lmsg("Exiting");
    exit(0);
}

int mkdir_all(char *path)
{
    char *p = path;
    while (*p != '\0')
    {
        if (*p == '/')
        {
            *p = '\0';
            mkdir(path, 0755);
            *p = '/';
        }
        p++;
    }
    return 0;
}

int tty_recv_fd(char *sock_path)
{
    sock_fd = socket(AF_UNIX, SOCK_STREAM, 0);
    if (sock_fd < 0)
    {
        return sock_fd;
    }
    lmsg("created socket");

    unlink(sock_path);

    char *dir = strdup(sock_path);
    int err = mkdir_all(dirname(dir));
    free(dir);
    if (err < 0)
    {
        close(sock_fd);
        return err;
    }

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(struct sockaddr_un));
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, sock_path, sizeof(addr.sun_path) - 1);

    lmsg("binding tty socket path");
    lmsg(sock_path);

    err = bind(sock_fd, (struct sockaddr *)&addr, sizeof(struct sockaddr_un));
    if (err < 0)
    {
        close(sock_fd);
        return err;
    }

    lmsg("bound socket");

    err = listen(sock_fd, 1);
    if (err < 0)
    {
        close(sock_fd);
        return err;
    }

    lmsg("listening on socket");

    err = sd_notify("READY=1");
    if (err < 0)
    {
        lerror("sd_notify");
    }
    else
    {
        lmsg("sd_notify READY=1");
    }

    int conn = accept(sock_fd, NULL, NULL);
    if (conn < 0)
    {
        close(sock_fd);
        return conn;
    }

    lmsg("accepted connection");

    char buf[512];
    struct iovec e = {buf, 512};
    char cmsg[CMSG_SPACE(sizeof(int))];
    struct msghdr m = {NULL, 0, &e, 1, cmsg, sizeof(cmsg), 0};

    int n = recvmsg(conn, &m, 0);
    if (n < 0)
    {
        close(conn);
        close(sock_fd);
        return n;
    }

    lmsg("received fd");

    struct cmsghdr *c = CMSG_FIRSTHDR(&m);
    int fd = *(int *)CMSG_DATA(c);

    close(conn);
    // We re-use sock_fd for the tty operations, so don't close that.
    return fd;
}

void pty_main(void)
{
    char *val = getenv("_TTY_HANDSHAKE");
    if (val == NULL || *val != '1')
    {
        return;
    }

    char *sock_path = getenv("_TTY_SOCKET_PATH");
    if (sock_path == NULL)
    {
        lerror("no socket path");
        exit(1);
    }

    setcgroup();
    lmsg("cgroup set");

    tty_fd = tty_recv_fd(sock_path);
    if (tty_fd < 0)
    {
        lerror("tty_recv_fd");
        exit(2);
    }

    // TODO: make this configurable
    // Maybe we can use the container uid by default (unless it is root)?
    if (setuid(10000) < 0)
    {
        lerror("setuid");
        exit(1);
    }

    if (handle_pty() < 0)
    {
        lerror("handle_pty");
        exit(3);
    }
}
